torch/nn/parallel/distributed_c10d.py

```diff
@@ -0,0 +1,352 @@
+
+        # module buffer sync
+        if self.broadcast_buffers:
+            buffers = list(self.module._all_buffers())
+            if len(buffers) > 0:
+                # cross-node buffer sync
+                self._dist_broadcast_coalesced(buffers, self.broadcast_bucket_size)
+
+                if len(self.device_ids) > 1:
+                    # intra-node buffer sync
+                    result = broadcast_coalesced(buffers, self.device_ids, self.broadcast_bucket_size)
+                    for tensors, module in zip(result[1:], self._module_copies[1:]):
+                        for tensor, buf in zip(tensors, module._all_buffers()):
+                            buf.set_(tensor)
+
+    def _register_grad_hooks(self):
+        self._grad_accs = []  # need to keep them in scope
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                if p.requires_grad:
+                    p_tmp = p.expand_as(p)
+                    grad_acc = p_tmp.grad_fn.next_functions[0][0]
+                    grad_acc.register_hook(self._make_param_hook(p, device_idx))
+                    self._grad_accs.append(grad_acc)
+
+    def _make_param_hook(self, param, device_idx):
+        bucket_idx = self.bucket_map[param]
+
+        def distributed_data_parallel_hook(*unused):
+            if param.grad.requires_grad:
+                raise RuntimeError("DistributedDataParallelC10d only works "
+                                   "with gradients that don't require grad")
+            bucket = self.buckets[bucket_idx][device_idx]
+            bucket.append(param.grad.data)
+
+            # We can flush these and save memory for replicas
+            if device_idx > 0:
+                param.grad = None
+                param.data.set_()
+
+            # Current device's bucket is full
+            if len(bucket) == self.bucket_sizes[bucket_idx]:
+                self.devs_ready[bucket_idx] += 1
+                if self.devs_ready[bucket_idx] < len(self.device_ids):
+                    return
+
+                if bucket_idx not in self.buckets_ready:
+                    self.buckets_ready.add(bucket_idx)
+                    if len(self.buckets_to_reduce) > 0 and self.buckets_to_reduce[0] == bucket_idx:
+                        # Reduce the bucket
+                        self._queue_reduction(bucket_idx)
+                        self.buckets_to_reduce.pop(0)
+
+                    # If all buckets are ready
+                    if len(self.buckets_ready) == len(self.bucket_order):
+                        # In case there are unreduced buckets, reduce them all
+                        if len(self.buckets_to_reduce) > 0:
+                            for b_idx in self.buckets_to_reduce:
+                                self._queue_reduction(b_idx)
+                        # A final sync for all the reduction works
+                        self._sync_reduction_works()
```

Comment:
This strategy seems a bit weird to me. Do I understand correctly, that basically if the buckets happen to be ready in `bucket_order`, then we'll be processing them in background as `backward` runs, but if there are any that are lagging behind (and this is possible due to no guarantees about ordering from autograd), then they will prevent the reduction of *any* of the buckets that come after them, until the very end of backward?

It seems to me like it would be simpler to:
1. If it's not the time for our bucket to reduce, then bail out (we got there too early)
2. Otherwise enter a for loop that starts at our bucket, and queues as many reduction as there are ready.

This will ensure we do as much parallel work as we possibly can, without waiting

Reply:
Yes, your understanding is correct.  We do need each process to have the exact order of reduction. With your proposal, won't it break this asssumption